backend/inspirehep/records/api/literature.py

```diff
@@ -267,106 +269,161 @@ def add_files(self, data):
-                new_file_data = self.add_file(**document)
-                document.update(new_file_data)
-            try:
-                builder.add_document(**document)
-            except ValueError:
-                LOGGER.exception(
-                    "Duplicated document found",
-                    recid=self.get("control_number"),
-                    uuid=self.id,
-                    document=document.get("key"),
+    def _generate_tasks(self, files, executor, add_file_func):
+        tasks = {}
+        for idx, file_data in enumerate(files):
+            if not file_data.get("hidden", False):
+                tasks.update(
+                    {
+                        executor.submit(
+                            self.add_file,
+                            current_app.app_context(),
+                            control_number=self.get("control_number"),
+                            uuid=self.id,
+                            **file_data,
+                        ): idx
+                    }
                 )
-        return builder.record.get("documents")
+            else:
+                self._update_file_entry(file_data, add_file_func)
 
-    def add_figures(self, figures):
-        builder = LiteratureBuilder()
-        for figure in figures:
-            new_file_data = self.add_file(**figure)
-            figure.update(new_file_data)
-            try:
-                builder.add_figure(**figure)
-            except ValueError:
-                LOGGER.exception(
-                    "Duplicated figure found",
-                    recid=self.get("control_number"),
-                    uuid=self.id,
-                    figure=figure.get("key"),
-                )
-        return builder.record.get("figures")
+        return tasks
 
-    def add_file(
-        self, url, original_url=None, key=None, filename=None, *args, **kwargs
-    ):
-        if current_s3_instance.is_s3_url(url) and not current_app.config.get(
-            "UPDATE_S3_FILES_METADATA", False
-        ):
-            result = {}
-            if key not in url:
-                filename = filename or key
-                key = url.split("/")[-1]
-                result.update({"key": key, "filename": filename})
-            LOGGER.info(
-                "File already on S3 - Skipping",
-                url=url,
-                key=key,
-                recid=self.get("control_number"),
-                uuid=self.id,
-            )
-            return result
-        file_data = download_file_from_url(url)
-        new_key = hash_data(file_data)
-        mimetype = magic.from_buffer(file_data, mime=True)
-        file_data = BytesIO(file_data)
-        filename = filename or key
-        acl = current_app.config["S3_FILE_ACL"]
-        if current_s3_instance.file_exists(new_key):
-            LOGGER.info(
-                "Replacing file metadata",
-                key=new_key,
+    def _update_file_entry(self, file_data, add_file_func):
```

Comment:
Why we need a serate function to just update the file? Why we need a serate function to just update the file? It was much nicer to keep figures  and documents logic seprated

Reply:
We use the same thing in 2 separate places. With the same try/except. It's just logical to keep it as separate function than have identical code in 2 places